### PR TITLE
Introduce pnpm lockfile v9 detector

### PR DESCRIPTION
**Pull request by FernandoRojo**
There is still no official docs for the new v9 lock if format, so this parsing logic is empirically based and will be updated as specs are released
Issue tracking v9 specs: https://github.com/pnpm/spec/issues/6
Format should eventually get updated here: https://github.com/pnpm/spec/blob/master/lockfile

Includes a small refactoring to re-use code and establish consistency between lockfile parsing.

### CODE REVIEW

File: test/Microsoft.ComponentDetection.Detectors.Tests/PnpmParsingUtilitiesTest.cs [hunk 1]
```diff
@@ -142,40 +145,44 @@ public void DeserializePnpmYamlFileV6()
...
+        pnpmParsingUtilities.ReconstructPnpmDependencyPath("@babel/eslint-parser", "7.23.3").Should().BeEquivalentTo("/@babel/eslint-parser@7.23.3");
 
         // With peer deps:
-        PnpmParsingUtilities.ReconstructPnpmDependencyPathV6("mocha-json-output-reporter", "2.1.0(mocha@10.2.0)(moment@2.29.4)").Should().BeEquivalentTo("/mocha-json-output-reporter@2.1.0(mocha@10.2.0)(moment@2.29.4)");
+        pnpmParsingUtilities.ReconstructPnpmDependencyPath("mocha-json-output-reporter", "2.1.0(mocha@10.2.0)(moment@2.29.4)").Should().BeEquivalentTo("/mocha-json-output-reporter@2.1.0(mocha@10.2.0)(moment@2.29.4)");
 
         // Absolute path:
-        PnpmParsingUtilities.ReconstructPnpmDependencyPathV6("events_pkg", "/events@3.3.0").Should().BeEquivalentTo("/events@3.3.0");
+        pnpmParsingUtilities.ReconstructPnpmDependencyPath("events_pkg", "/events@3.3.0").Should().BeEquivalentTo("/events@3.3.0");
     }
```
Comment (FernandoRojo) [inline test/Microsoft.ComponentDetection.Detectors.Tests/PnpmParsingUtilitiesTest.cs:187]: Add parsing tests for v9 lockfiles

**Review by grvillic — APPROVED** [review]

File: src/Microsoft.ComponentDetection.Detectors/pnpm/Pnpm9Detector.cs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        var components = new Dictionary<string, (DetectedComponent, Package)>();
+
+        // Create a component for every package referenced in the lock file.
+        // This includes all directly and transitively referenced dependencies.
+        foreach (var (pnpmDependencyPath, package) in yaml.Snapshots ?? Enumerable.Empty<KeyValuePair<string, Package>>())
+        {
+            // Ignore "file:" as these are local packages.
+            // Such local packages should only be referenced at the top level (via ProcessDependencyList) which also skips them or from other local packages (which this skips).
+            // There should be no cases where a non-local package references a local package, so skipping them here should not result in failed lookups below when adding all the graph references.
+            if (pnpmDependencyPath.StartsWith("file:"))
```
Comment (grvillic) [inline src/Microsoft.ComponentDetection.Detectors/pnpm/Pnpm9Detector.cs:33]: nit: we can create a constant for this pnpm Dependency Path and share across all the versions that need it

File: src/Microsoft.ComponentDetection.Detectors/pnpm/ParsingUtilities/PnpmV9ParsingUtilities.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+        // Version is section after last `@`.
+        var packageVersion = packageNameParts[^1];
+
+        return new DetectedComponent(new NpmComponent(fullPackageName, packageVersion));
+    }
+
+    public override bool IsPnpmPackageDevDependency(Package pnpmPackage)
+    {
+        // TODO: Implement
```
Comment (grvillic) [inline src/Microsoft.ComponentDetection.Detectors/pnpm/ParsingUtilities/PnpmV9ParsingUtilities.cs:34]: Is this completely missing or are we capturing dev dependencies in a different way for V9?

Comment (FernandoRojo) [reply src/Microsoft.ComponentDetection.Detectors/pnpm/ParsingUtilities/PnpmV9ParsingUtilities.cs:34]: It's completely missing, dev dependency resolution in the new v9 detector is manually determined when traversing the dependency tree

**Pull request merged**